File: frontend/services/leaderboard.ts
import { api } from "./api";

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  username: string;
  net_worth: number;
  investor_rank: number;
  total_cards_played: number;
  portfolio_id: string;
}

export interface LeaderboardData {
  entries: LeaderboardEntry[];
  current_user_rank: number | null;
}

const RANK_LABELS = ["", "ANALYST", "ASSOCIATE", "VP", "MD"];

export async function getLeaderboard(limit: number = 50): Promise<LeaderboardData> {
  const response = await api.get<LeaderboardData>("/api/leaderboard", {
    params: { limit },
  });
  return response.data;
}

export function getRankLabel(rank: number): string {
  return RANK_LABELS[Math.min(rank, RANK_LABELS.length - 1)] || "MD";
}

File: backend/app/schemas/leaderboard.py
from pydantic import BaseModel


class LeaderboardEntryOut(BaseModel):
    """A single entry in the leaderboard"""

    rank: int
    user_id: str
    username: str
    net_worth: float
    investor_rank: int
    total_cards_played: int
    portfolio_id: str


class LeaderboardOut(BaseModel):
    """The full leaderboard response"""

    entries: list[LeaderboardEntryOut]
    current_user_rank: int | None
